plugins/woocommerce/src/Internal/Admin/Features/ProductBlockEditor/ProductFormTemplates/BlockContainerTrait.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductFormTemplates;
+
+use Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductFormTemplates\BlockInterface;
+
+/**
+ * Trait for block containers.
+ */
+trait BlockContainerTrait {
+	/**
+	 * The child blocks.
+	 *
+	 * @var BlockInterface[]
+	 */
+	private $child_blocks = [];
+
+	/**
+	 * Add a block to the block container.
+	 *
+	 * @param array $block_config The block data.
+	 *
+	 * @throws \ValueError If the block configuration is invalid.
+	 * @throws \ValueError If a block with the specified ID already exists in the template.
+	 */
+	public function &add_block( array $block_config ): BlockInterface {
```

Comment:
My thought was that

1. More specialized template interfaces will implement more specialized methods for getting/setting...

```
interface ProductFormTemplate {
    public function add_group(): GroupInterface;

    public function get_group(): GroupInterface;
    public function get_section(): SectionInterface;
    public function get_field(): FieldInterface;
}

interface GroupInterface {
    public function add_section(): SectionInterface;
}
```

And, any extension points (hooks, etc.) should always use interfaces, never classes. So, only what is exposed on the interface will be available.

If an implementation really wanted to expose BlockContainerInterface (as opposed to creating their own interface) but not support `add_block`, they could override the `add_block` implementation to throw a 'not support' exception.

Reply:
Just to make sure I understand, we never plan to directly call `use BlockContainerTrait;` or extend from `BlockTemplate` but instead implement specific methods and interface for customized templates.  Does that sound correct?

So internally, the product form template might do something like this, using `Block` internally but not exposing its methods:

```php
class ProductFormTemplate implements ProductFormTemplate {
    public function add_group( $args ) {
        $block = new Block( ... );
        $root_template->internal_add_block_to_template( $block );
    }
}
```

My only concern with this is that there's a lot of boilerplate for templates and blocks to reimplement.  Presumably if we had a `Section` type of of block, it would want access to `get_id`, `get_order`, all the properties, etc.  This also has potential to create bugs since we can't reliably know if customized "block" classes have certain methods.

But I might be misunderstanding your idea here, so feel free to steer me in the right direction if so.

Reply:
This is roughly what I'm thinking...

In this example...

* a `Group` cannot have blocks added directly to it, only `Sections`.
* a `ProductFormTemplate` cannot have blocks added directly to it, only `Groups`.

```php
interface GroupInterface extends BlockInterface {
    public function add_section( array $block_config ): SectionInterface;
}

class Group extends Block implements GroupInterface {
    public function __construct( array $config, BlockTemplateInterface &$root_template, BlockContainerInterface &$parent = null ) {
        parent::__construct( $config, $root_template, $parent );
    }    

    public function add_block( array $block_config ): BlockInterface {
        // see below for similar implementation 
    }    

    public function add_section( array $block_config ): SectionInterface {
        // see below for similar implementation
    }
}

interface ProductFormTemplateInterface extends BlockTemplateInterface {
    public function add_group(): GroupInterface;

    public function get_group( string $id ): GroupInterface;
}

class ProductFormTemplate extends BlockTemplate implements ProductFormTemplateInterface {
    public function add_block( array $block_config ): BlockInterface {
        // there might be a better built-in exception class,
        // or we could create a custom one; the basic idea is
        // to throw an exception to make it clear we don't support this method
        // (this would be documented in code docs and other places of course)
        throw new BadMethodCallException();
    }    

    // I added the ability to pass in a custom block generator to allow for more reuse
    public function add_group( array $block_config ): GroupInterface {
        $group = parent::add_block(
            $block_config,
            function  ( array $config, BlockTemplateInterface &$root_template, BlockContainerInterface &$parent = null ) {
                return new CustomBlock( $config, $root_template, $parent );
            }
        );
    }

   public function &get_group( string $id ): GroupInterface {
       $block = parent::get_block( $id );
       if ( isset( $block ) && ! $block instanceof GroupInterface ) {
           throw new TypeError();
       }
       return $block;
   }
}
```

Reply:
I pushed a commit with the add_block change to allow for a custom block generator... will fix up checks tomorrow morning.

Reply:
I've added tests around this change to `add_block` to allow for a "block creator" to be passed in, which allows for custom "sub-types" of `BlockInterface` to be added, without much boilerplate at all.

I think with this, it will be easy for custom template implementations to provide a more restrictive interface without having to re-implement any of the core processing that we provide in the base implementation.

And, keep in mind that in general this sort of custom template implementation really would be internal to WooCommerce core for the most part. 3rd parties would not need to do any of this to simply modify a template to get their blocks added to it.

Reply:
I think in general that overriding methods in subclasses and throwing warnings may be a code smell and mean that the subclass is not truly representative of the parent or that the parent class is overreaching and doing more than it should.  In this case, I think it's the latter.

What if we simplify this and break apart the pieces that are true of all blocks and create the more distinguishable features as traits.

```php
class Block {
    public function get_id()
    public function get_order()
    // other common block functionality that should exist in all blocks
}

trait BlockContainerTrait {
    public function add_block()
}

// Optional generic block container that allows inner blocks
class BlockContainer extends Block implements BlockInterface, BlockContainerInterface {
    use BlockContainerTrait;
}
```

Then more custom blocks can compose themselves using relevant traits.

```php
trait GroupBlockTrait {
    public function add_section()
}

class GroupBlock extends Block {
    use GroupBlockTrait;
}

class CustomGroupBlock extends Block {
    use GroupBlockTrait;
}
```

I think this makes more sense since we can say that not all blocks should even be block containers and allow children.